# ユーザー定義関数／クラス／メソッド／モジュール

---

- ref: https://docs.python.org/ja/3/tutorial/classes.html
- blog: https://slash-z.com/
- github: https://github.com/KazutoMakino/PythonCourses

---

## 演習問題の解答

- Q1. 数値のリスト 1 つを仮引数とし，戻り値をこのリストの RMS 値 1 つとした関数: `calc_rms` を作成してください．ここで，データの個数を $n$, i番目のデータを $x_i$ とすると，RMS 値は，

$$
RMS = \sqrt{\frac{1}{n} \sum_{i=1}^{n}x_i^2}
$$

で計算されます．

In [41]:
def calc_rms(arr):
    # init
    ret = 0
    
    for x in arr:
        # add squared value
        ret += x**2
    
    # mean
    ret /= len(arr)
    
    # sqrt
    ret = ret**(1/2)
    
    return ret

または，

In [42]:
def calc_rms(arr):
    return (sum([v**2 for v in arr]) / len(arr))**(1/2)

- Q2. Q1 で作成した `calc_rms` の引数に 1 ～ 100 の奇数のリストを入力し，結果（戻り値）が 57.73214... であることを確認してください．

In [43]:
arr = list(range(1,101,2))
ret = calc_rms(arr=arr)
ret, round(ret, 5)==57.73214

(57.73214009544424, True)

- Q3. `./data.txt` を読み，戻り値が `Hmidity[%]` の値のみのタプルとした引数なしの関数: `load_humi_data` を作成してください．データの読み込みにおいて，`import` が必要な標準関数やライブラリを用いず，テキストデータとして読み込んだものを処理する関数にしてください．

まずは，ファイルの中身を見てみます．  
読み込んだ際に全て表示すると大きい場合があるので，読み込んだら，文字数をまず出力してみます．

In [44]:
with open("./data.txt", mode="r") as f:
    txt = f.read()
len(txt)

59462

59462 文字ということで，全部表示させると邪魔なので，前半／後半のそれぞれ 500 文字について表示してみます．

In [45]:
txt[:500]

'TimeStamp: 2022/01/18 19:29:14.688684, ElapsedTime[s]: 114.311, Temperature[degC]: 19.40032, Humidity[%]: 56.44159\nTimeStamp: 2022/01/18 19:29:16.22105, ElapsedTime[s]: 115.644, Temperature[degC]: 19.38697, Humidity[%]: 55.97925\nTimeStamp: 2022/01/18 19:29:17.355098, ElapsedTime[s]: 116.977, Temperature[degC]: 19.40032, Humidity[%]: 55.46502\nTimeStamp: 2022/01/18 19:29:18.688375, ElapsedTime[s]: 118.31, Temperature[degC]: 19.40032, Humidity[%]: 54.99352\nTimeStamp: 2022/01/18 19:29:20.21160, Elap'

In [46]:
txt[-500:]

'e[degC]: 19.6273, Humidity[%]: 46.95049\nTimeStamp: 2022/01/18 19:40:42.663524, ElapsedTime[s]: 802.285, Temperature[degC]: 19.61395, Humidity[%]: 46.89556\nTimeStamp: 2022/01/18 19:40:43.996179, ElapsedTime[s]: 803.618, Temperature[degC]: 19.60059, Humidity[%]: 46.91997\nTimeStamp: 2022/01/18 19:40:45.329131, ElapsedTime[s]: 804.951, Temperature[degC]: 19.61395, Humidity[%]: 46.92302\nTimeStamp: 2022/01/18 19:40:46.660690, ElapsedTime[s]: 806.282, Temperature[degC]: 19.60059, Humidity[%]: 46.86809\n'

ということでデータ形式としては辞書のような形で，変数名と値がコロンで 1 対 1 に対応し，一つのタイムスタンプに対してカンマ区切りでデータが 3 つあり，次のタイムスタンプとデータの登録は `\n` による改行で区切られています（もし，改行コードが `\n` 以外に書き換えられていたら，そちらに読み替えてください）．  
題意の `Humidity[%]` の値のみのリストを取得するためには，

- 改行 `\n` で区切ったリストを作る
- 上記の各要素について `Humidity[%]: ` を含むこれ以前の文字列を削除し，float でキャストする

上記 2 つの工程を経て実現できると考えられますので，以下のように実装します．

In [52]:
def load_humi_data(fpath="./data.txt"):
    # laod data.txt
    with open(fpath, mode="r") as f:
        txt = f.read()
    
    # split: "\n"
    data = txt.split("\n")
    
    # cut and cast to float
    data = tuple(float(v.split("Humidity[%]: ")[-1]) for v in data if v)
    
    return data

- Q4. Q3 の戻り値であるタプルについて，先頭／末尾それぞれ 5 つを表示し，それぞれ `(56.44159, 55.97925, 55.46502, 54.99352, 54.72496)` と `(46.95049, 46.89556, 46.91997, 46.92302, 46.86809)` であることと，タプルの中身の要素の一つを取り出し，その型が `float` であることを確認ください．

In [54]:
data = load_humi_data()
data[:5]

(56.44159, 55.97925, 55.46502, 54.99352, 54.72496)

In [55]:
data[-5:]

(46.95049, 46.89556, 46.91997, 46.92302, 46.86809)

In [56]:
type(data[0])

float

- Q5. 以下のクラス定義／メソッド実行を行ってください（結果が合えば良く，細部の差異については気にする必要はありません）．
    - Q1 と Q3 で作成した `calc_rms` と `load_humi_data` をメソッドに持つクラス: `HumidityRMSGetter` を作成してください．
    - 時系列データである `load_humi_data` の戻り値のタプルについて，そのデータ点を含んだ時間方向に対してそれ以前の計 10 点のデータに対する RMS 値を格納したリストを返すメソッド: `get_rms_arr` を作成してください．このとき，`get_rms_arr` に渡す配列の最初の方においてデータ点は 10 点に届きませんが，この場合はそのデータ点含むそれ以前のデータ点全てとしてください．また，`get_rms_arr` にて返す RMS 値は，小数点第5位までとし，小数点第6位に偶数丸めを適用したものとしてください．
    - 最後に，`HumidityRMSGetter` をインスタンス化し，`get_rms_arr` を呼び出して結果を変数に代入し，先頭／末尾それぞれ 3 つ表示し，それぞれの結果が `[56.44159, 56.2109, 55.96337]` と `[46.87241, 46.8774, 46.87726]` であることを確認ください．

In [71]:
class HumidityRMSGetter:
    def __init__(self, fpath="./data.txt", smp_rate=10):
        self.fpath = fpath
        self.smp_rate = smp_rate
        
    @staticmethod
    def calc_rms(arr):
        return (sum([v**2 for v in arr]) / len(arr))**(1/2)

    def load_humi_data(self):
        # laod data.txt
        with open(self.fpath, mode="r") as f:
            txt = f.read()

        # split: "\n"
        self.data = txt.split("\n")

        # cut and cast to float
        self.data = tuple(
            float(v.split("Humidity[%]: ")[-1]) for v in self.data if v
        )
        
    def get_rms_arr(self):
        # get humidity tuple
        self.load_humi_data()
        
        # init
        rms_arr = []
        
        # loop: range(len(self.data))
        for i in range(len(self.data)):
            # get data into calc_rms
            data = self.data[:i + 1] if i < self.smp_rate else self.data[i - self.smp_rate : i + 1]
                
            # calc rms
            rms = HumidityRMSGetter.calc_rms(arr=data)
            
            # round
            rms = round(rms, 5)
            
            # append
            rms_arr.append(rms)
            
        return rms_arr
    

# instance
hum_rms_getter = HumidityRMSGetter()

# get rms list
rms_arr = hum_rms_getter.get_rms_arr()

In [72]:
rms_arr[:3]

[56.44159, 56.2109, 55.96337]

In [73]:
rms_arr[-3:]

[46.87241, 46.8774, 46.87726]